### PR TITLE
InvalidOperationException thrown while removing the subscribers in delete persistent subscription process. Modify the update persistent subscription strategy

### PR DESCRIPTION
**Pull request by dhingrak**
Fixed:  #3512  duplicate PS in the EventStoreDB UI

### CODE REVIEW

**Comment by shaan1337** [comment]
hey @dhingrak,

just wanting to make sure i understand the problem and solution correctly:

- my (rough) understanding after reading #3512 , is that the root cause is that saving the configuration to the config stream is done asynchronously and this opens up the possibility for a persistent subscription to be deleted in the _subscription* dictionaries but not yet in the config stream - so creating a new one could result in a duplicate. however, what's not clear to me, is that even if the above scenario occurred, _config should not be able to contain two subscriptions with the same name because the messages handled in the persistent subscription service are handled one at a time (see comment below about concurrency in solution section).

- to reproduce the issue, how did you simulate step 2 in #3512 ?
 
- we also have the issue where deleting/cleaning up the checkpoint and parked messages streams are done asynchronously and it may never actually happen if there is a timeout (but that's another problem)

solution:
- based on the code changes, it seems that you're relying on a lock to prevent a concurrent update (if i'm understanding it correctly). however a lock shouldn't be required since all the messages handled by the persistent subscriptions service are serialized in a queue and executed one at a time.

**Comment by dhingrak** [comment]
@shaan1337 
Thanks a lot Shaan for your comments. Actually, I forgot to update the problem statement for https://github.com/EventStore/EventStore/issues/3512. 
After more investigation and discussion with Hayley we figured out it could be the create subscription part that are not handling the simultaneously request from clients.

I have tested the below scenario in my local environment. To reproduce this issue I have added a delay of 500 ms in createSubscriptionGroup section code and am able to create duplicate subscriptions. By adding a delay the check for existing persistent subscriptions passes when there are simultaneous requests and we have a duplicate subscription in UI. 

I have added a lock on the _subscriptionsById so the two clients are not able to update the state of _subscriptiionsById simultaneously to avoid the duplicate PS creation. 

**Update persistent subscription changes:** 

I have modify the code as suggested by Hayley so the update PS process completes in a single step rather than deleting the PS and then create a new PS

**Comment by shaan1337** [comment]
@dhingrak thanks for all the clarifications :+1: i'll give it a try

**Comment by shaan1337** [comment]
hey @dhingrak, i've checked the repro code you've posted in https://github.com/EventStore/EventStore/issues/3512

by adding `Task.Delay(...).ContinueWith...`, it is possible that the continuation will be run on another thread, which would cause the PersistentSubscriptionService to longer be thread-safe and could subsequently allow creation of duplicate persistent subscriptions.

would you be able to try with `Thread.Sleep(1000);` instead?

**Comment by dhingrak** [comment]
hey @shaan1337 

I have tested the scenario with Thread.Sleep(1000) and I am not able to create duplicate PS. Indeed, the `Task.Delay(...).ContinueWith..` running on another thread in my local testing and that's why there are duplicate PS

**Comment by dhingrak** [comment]
The above PR fixes the issue for InvalidOperationException thrown during the subscriber.shutdown process. Also, as suggested by Hayley we have modify the logic for updating the persistent subscriptions. Currently, the update process consists of delete and then recreate the persistent subscription. This PR will update the persistent subscription without deleting and recreating it.

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 1]
```diff
@@ -399,18 +402,19 @@ string user
...
 		) {
 			if (!_started) return;
-			var key = BuildSubscriptionGroupKey(eventSource.ToString(), groupName);
+			var stream = eventSource.ToString();
+			var key = BuildSubscriptionGroupKey(stream, groupName);
 			Log.Debug("Updating persistent subscription {subscriptionKey}", key);
 
 			if (!_subscriptionsById.ContainsKey(key)) {
-				onNotExist($"Group '{groupName}' does not exist.");
+				onNotExist($"Group '{groupName}' does not exists.");
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:414]:
```suggestion
				onNotExist($"Group '{groupName}' does not exist.");
```

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 2]
```diff
@@ -626,22 +675,45 @@ string user
...
 			Log.Debug("Deleting persistent subscription {subscriptionKey}", key);
 
 			PersistentSubscription subscription;
+			
+			List<PersistentSubscription> subscribers;
+
+
 			if (!_subscriptionsById.TryGetValue(key, out subscription)) {
-				onNotExist($"Group '{groupName}' does not exist.");
+				onNotExist($"Group '{groupName}' does not exists.");
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:688]:
```suggestion
				onNotExist($"Group '{groupName}' does not exist.");
```

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 3]
```diff
@@ -399,18 +402,19 @@ string user
...
+			var key = BuildSubscriptionGroupKey(stream, groupName);
 			Log.Debug("Updating persistent subscription {subscriptionKey}", key);
 
 			if (!_subscriptionsById.ContainsKey(key)) {
-				onNotExist($"Group '{groupName}' does not exist.");
+				onNotExist($"Group '{groupName}' does not exists.");
 				return;
 			}
-
+
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:417]:
```suggestion
```

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 4]
```diff
@@ -612,8 +647,22 @@ string user
 					_checkpointReader,
 					new PersistentSubscriptionCheckpointWriter(key, _ioDispatcher),
 					new PersistentSubscriptionMessageParker(key, _ioDispatcher)));
+			
+			if (_subscriptionsById.ContainsKey(key)) {
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:651]: suggestion: move this check before creating the PersistentSubscription object, otherwise it's not used.
it should also be ok to remove the Log.Information line since we don't currently log these errors:

```
			if (_subscriptionsById.ContainsKey(key))
				return false;
```

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 5]
```diff
@@ -612,8 +647,22 @@ string user
...
+				Log.Information("GroupName already exists: " + groupName);
+				return false;
+			}
+
 			_subscriptionsById[key] = subscription;
+
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:657]: suggestion:
move `List<PersistentSubscription> subscribers;` down here

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 6]
```diff
@@ -420,31 +424,55 @@ string user
...
 
-			_config.Updated = DateTime.Now;
-			_config.UpdatedBy = user;
-			_config.Entries.Add(new PersistentSubscriptionEntry {
-				Stream = eventSource.ToString(), //'Stream' name kept for backward compatibility
+						subscribers[i] = subscription;
+						break;
+					}
+				}
+			} else {
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:468]: it seems that `subscription` is not being added to the `subscribers` list in the `else` part

actually, should we throw an exception in the `else` part? since the subscription should exist in `_subscriptionTopics` at this point.

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs (hunk 1 above)
Comment (dhingrak) [reply src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:414]: Done

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs (hunk 2 above)
Comment (dhingrak) [reply src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:688]: Done

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs (hunk 4 above)
Comment (dhingrak) [reply src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:651]: Made the changes

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs (hunk 5 above)
Comment (dhingrak) [reply src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:657]: Done

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 7]
```diff
@@ -719,29 +791,6 @@ string user
 			);
 		}
 
-		private void RemoveSubscriptionConfig(string username, string eventSource, string groupName) {
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:722]: suggestion:
to avoid code repetition in the update & remove methods, could we keep these two methods and add a parameter named: `replaceBy`  ?

```
RemoveSubscriptionConfig(string username, string eventSource, string groupName, PersistentSubscriptionEntry replaceBy);
RemoveSubscription(string eventSource, string groupName, PersistentSubscription replaceBy)
```

when `replaceBy` is null, we call RemoveAt(index) and when it's not null, we can replace the value at the index with that subscription or config.

Comment (shaan1337) [reply src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:722]: the replacement null check would also need to be done for `_subscriptionsById.Remove(key);`

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs (hunk 6 above)
Comment (shaan1337) [reply src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:468]: edit: this may no longer be relevant if we go for the change suggested here: https://github.com/EventStore/EventStore/pull/3533/files#r987722268

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs (hunk 7 above)
Comment (shaan1337) [reply src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:722]: for more clarity, maybe we could then rename `RemoveSubscriptionConfig` to `UpdateSubscriptionConfig` and `RemoveSubscription` to `UpdateSubscription`

Comment (dhingrak) [reply src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:722]: Thanks a lot Shaan for the suggestions. Changes updated

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 8]
```diff
@@ -460,10 +463,15 @@ string user
...
-				StartFrom = startFrom is PersistentSubscriptionSingleStreamPosition x ? x.StreamEventNumber : long.MinValue,
+				StartFrom = startFrom is PersistentSubscriptionSingleStreamPosition x
+					? x.StreamEventNumber
+					: long.MinValue,
 				#pragma warning restore 612
 				StartPosition = startFrom.ToString()
-			});
+			};
+			
+			UpdateSubscription(stream,groupName,subscription);
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:473]:
```suggestion
			UpdateSubscription(stream, groupName, subscription);
```

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 9]
```diff
@@ -718,24 +737,47 @@ string user
...
 			_config.UpdatedBy = username;
 			var index = _config.Entries.FindLastIndex(x => x.Stream == eventSource && x.Group == groupName);
-			_config.Entries.RemoveAt(index);
+			
+			// Updating the persistent subscription if replaceBy is not null, otherwise delete the persistent subscription
+			if (replaceBy != null) {
+				_config.Entries[index] = replaceBy;
+			} else {
+				_config.Entries.RemoveAt(index);
+			}
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:751]:
```suggestion
			// Update the persistent subscription if replaceBy is not null, otherwise delete the persistent subscription
			if (replaceBy != null)
				_config.Entries[index] = replaceBy;
			else
				_config.Entries.RemoveAt(index);
```

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 10]
```diff
@@ -718,24 +737,47 @@ string user
...
 			var key = BuildSubscriptionGroupKey(eventSource, groupName);
-			_subscriptionsById.Remove(key);
+			
+			// Updating the persistent subscription if replaceBy is not null, otherwise delete the persistent subscription
+			if(replaceBy != null)
+			{
+				_subscriptionsById[key] = replaceBy;
+			} else {
+				_subscriptionsById.Remove(key);
+			}
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:764]:
```suggestion
			// Update the persistent subscription if replaceBy is not null, otherwise delete the entry
			if (replaceBy != null)
				_subscriptionsById[key] = replaceBy;
			else
				_subscriptionsById.Remove(key);
```

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 11]
```diff
@@ -718,24 +737,47 @@ string user
...
+							sub.Shutdown();
+						} catch (Exception ex) {
+							Log.Error(ex, "Failed to shut down subscriber with id: {subscriptionId}", sub.SubscriptionId);
+						}
+
+						if (replaceBy != null) {
+							subscribers[i] = replaceBy;
+						} else {
+							subscribers.RemoveAt(i);
+						}
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:780]:
```suggestion
						if (replaceBy != null)
							subscribers[i] = replaceBy;
						else
							subscribers.RemoveAt(i);
```

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 12]
```diff
@@ -718,24 +737,47 @@ string user
...
 			if (_subscriptionTopics.TryGetValue(eventSource, out subscribers)) {
 				for (int i = 0; i < subscribers.Count; i++) {
 					var sub = subscribers[i];
 					if (sub.SubscriptionId == key) {
-						sub.Shutdown();
-						subscribers.RemoveAt(i);
+						try {
+							sub.Shutdown();
+						} catch (Exception ex) {
+							Log.Error(ex, "Failed to shut down subscriber with id: {subscriptionId}", sub.SubscriptionId);
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:773]:
```suggestion
							Log.Error(ex, "Failed to shut down subscription with id: {subscriptionId}", sub.SubscriptionId);
```

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 13]
```diff
@@ -399,18 +400,19 @@ string user
 				Action<string> onSuccess,
 				Action<string> onFail,
 				Action<string> onNotExist,
-				Action<string> onAccessDenied,
+				Action<string> onAccessDenied,
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:403]:
```suggestion
				Action<string> onAccessDenied,
```

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 14]
```diff
@@ -583,13 +590,13 @@ string user
...
-			if (!_subscriptionTopics.TryGetValue(eventSource.ToString(), out subscribers)) {
-				subscribers = new List<PersistentSubscription>();
-				_subscriptionTopics.Add(eventSource.ToString(), subscribers);
-			}
+			var stream = eventSource.ToString();
+			var key = BuildSubscriptionGroupKey(stream, groupName);
 
+			if (_subscriptionsById.ContainsKey(key)) {
+				return false;
+			}
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:598]:
```suggestion
			if (_subscriptionsById.ContainsKey(key))
				return false;
```

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 15]
```diff
@@ -253,11 +249,16 @@ string user
 				ToMessageTimeout(messageTimeoutMilliseconds)
 			);
 
+			if (!result) {
+				onExists($"Group '{groupName}' already exists.");
+				return;
+			}
+
 			Log.Debug("New persistent subscription {subscriptionKey}", key);
 			_config.Updated = DateTime.Now;
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:258]: similar to: https://github.com/EventStore/EventStore/pull/3533/files#r988645440
we can refactor this part and call `UpdateSubscriptionConfig(user, stream, groupName, createEntry);`

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs (hunk 7 above)
Comment (shaan1337) [reply src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:722]: thanks very much Kunal. I've made a few formatting suggestions and some additional refactoring suggestions for the creation part.

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs (hunk 1 above)
Comment (shaan1337) [reply src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs]: thanks!

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs (hunk 4 above)
Comment (shaan1337) [reply src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs]: thanks!

**Comment by dhingrak** [comment]
Thanks a lot Shaan for helping me in refactoring the code.

File: src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs [hunk 16]
```diff
@@ -235,8 +231,8 @@ string user
 				onFail(startFromValidationError);
 				return;
 			}
-
-			CreateSubscriptionGroup(eventSource,
+			
+			var result = CreateSubscriptionGroup(eventSource,
```
Comment (shaan1337) [inline src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs:235]: suggestion:
1. we can rename `CreateSubscriptionGroup` to `TryCreateSubscriptionGroup` 
2. [here](https://github.com/EventStore/EventStore/blob/0f4a2db987faeec93d642c2dcc6cf7d63b54837f/src/EventStore.Core/Services/PersistentSubscription/PersistentSubscriptionService.cs#L1028), we can log a warning when TryCreateSubscriptionGroup() returns false:
```
var key = BuildSubscriptionGroupKey(eventSource.ToString(), groupName);
Log.Warning("A duplicate persistent subscription: {subscriptionKey} was found in the configuration. Ignoring it.", key);
```

**Review by shaan1337 — APPROVED** [review]

**Review by thefringeninja — APPROVED** [review]
